src/mono/mono/metadata/image.c

```diff
@@ -1846,34 +1846,70 @@ mono_image_loaded_by_guid (const char *guid)
 	return mono_image_loaded_by_guid_internal (guid, FALSE);
 }
 
+#ifdef ENABLE_NETCORE
+static const char *
+mono_get_image_culture (MonoImage *image)
+{
+	MonoTableInfo *t = &image->tables [MONO_TABLE_ASSEMBLY];
+	if (!t->rows)
+		return NULL;
+
+	guint32 cols [MONO_ASSEMBLY_SIZE];
+	mono_metadata_decode_row (t, 0, cols, MONO_ASSEMBLY_SIZE);
+	return mono_metadata_string_heap (image, cols [MONO_ASSEMBLY_CULTURE]);
+}
+#endif
+
+char *
+mono_image_get_name_with_culture_if_needed (MonoImage *image)
+{
+#ifndef ENABLE_NETCORE
+	return NULL;
+#else
+	if (!g_str_has_prefix (image->name, "data-") &&
+		!g_path_is_absolute (image->name))
+	{
+		const char *culture = mono_get_image_culture (image);
+
+		if (culture && culture[0] != 0)
+			return g_strdup_printf ("(%s)%s", culture, image->name);
+	}
+
+	return NULL;
+#endif
+}
+
 static MonoImage *
 register_image (MonoLoadedImages *li, MonoImage *image, gboolean *problematic)
 {
 	MonoImage *image2;
+	char *name_with_culture = mono_image_get_name_with_culture_if_needed (image);
 	GHashTable *loaded_images = mono_loaded_images_get_hash (li, image->ref_only);
 
 	mono_images_lock ();
-	image2 = (MonoImage *)g_hash_table_lookup (loaded_images, image->name);
+	image2 = (MonoImage *)g_hash_table_lookup (loaded_images, name_with_culture != NULL ? name_with_culture : image->name);
```

Comment:
Rather than using the ternary operator a bunch, just set the name we intend to use to some variable and use that throughout.